[FLINK-30165][runtime][JUnit5 Migration] Migrate unaligned checkpoint related tests under flink-runtime module to junit5

1996fanrui opened the pull request:
## What is the purpose of the change

Migrate unaligned checkpoint related tests under flink-runtime module to junit5

## Brief change log

Migrate unaligned checkpoint related tests under flink-runtime module to junit5

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not documented

1996fanrui commented:
Hi @XComp , please help take a look in your free time, thanks!

XComp commented:
@snuyanzin may you have time to take a look at it?

1996fanrui commented:
Hi @RocMarshal  , please help take a look in your free time, thanks~

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateCheckpointWriterTest.java:
```diff
@@ -121,13 +122,15 @@ public void testSmallFilesNotWritten() throws Exception {
-        assertTrue(result.isDone());
-        assertEquals(0, checkpointsDir.list().length);
-        assertEquals(0, sharedStateDir.list().length);
+        assertThat(result.isDone()).isTrue();
+        String[] list = checkpointsDir.list();
+        int length = list.length;
+        assertThat(length).isEqualTo(0);
+        assertThat(sharedStateDir.list().length).isEqualTo(0);
```
```suggestion
        assertThat(checkpointsDir).isEmptyDirectory();
        assertThat(sharedStateDir).isEmptyDirectory();
```
Could be a bit simplified

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateCheckpointWriterTest.java:
```diff
@@ -218,12 +221,14 @@ public void testRecordingOffsets() throws Exception {
+                                    + numBytes * offsetCounts.remove(handle.getInfo()));
         }
-        assertTrue(offsetCounts.isEmpty());
+        assertThat(offsetCounts.isEmpty()).isTrue();
```
```suggestion
        assertThat(offsetCounts).isEmpty();
```
assertJ has `isEmpty()`

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateChunkReaderTest.java:
```diff
@@ -35,31 +35,44 @@
+                            } finally {
+                                checkState(serializer.failed);
+                                checkState(!handler.requestedBuffers.isEmpty());
+                                assertThat(
+                                                handler.requestedBuffers.stream()
+                                                        .allMatch(
+                                                                TestChannelStateByteBuffer
+                                                                        ::isRecycled))
+                                        .isTrue();
```
```suggestion
                                assertThat(handler.requestedBuffers.stream())
                                        .allMatch(TestChannelStateByteBuffer::isRecycled);
```
assertJ allows applying conditions to streams

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateChunkReaderTest.java:
```diff
@@ -35,31 +35,44 @@
+                                                "channelInfo",
+                                                0);
+                            } finally {
+                                checkState(serializer.failed);
+                                checkState(!handler.requestedBuffers.isEmpty());
```
```suggestion
                                assertThat(serializer.failed).isTrue();
                                assertThat(handler.requestedBuffers).isNotEmpty();
```
It seems `checkState` could be replaced with `assertThat` since the latest gives a better message

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateChunkReaderTest.java:
```diff
@@ -68,27 +81,28 @@ public void testBufferRecycledOnSuccess() throws IOException, InterruptedExcepti
                     .readChunk(stream, serializer.getHeaderLength(), handler, "channelInfo", 0);
         } finally {
             checkState(!handler.requestedBuffers.isEmpty());
```
```suggestion
            assertThat(handler.requestedBuffers).isNotEmpty();
```
It seems `checkState` could be replaced with `assertThat` since the latest gives a better message

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateChunkReaderTest.java:
```diff
@@ -68,27 +81,28 @@ public void testBufferRecycledOnSuccess() throws IOException, InterruptedExcepti
-            assertTrue(
-                    handler.requestedBuffers.stream()
-                            .allMatch(TestChannelStateByteBuffer::isRecycled));
+            assertThat(
+                            handler.requestedBuffers.stream()
+                                    .allMatch(TestChannelStateByteBuffer::isRecycled))
+                    .isTrue();
```
```suggestion
            assertThat(handler.requestedBuffers.stream())
                    .allMatch(TestChannelStateByteBuffer::isRecycled);
```
assertJ allows applying conditions to streams

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateChunkReaderTest.java:
```diff
@@ -68,27 +81,28 @@ public void testBufferRecycledOnSuccess() throws IOException, InterruptedExcepti
                     .readChunk(stream, serializer.getHeaderLength(), handler, "channelInfo", 0);
         } finally {
-            assertTrue(handler.requestedBuffers.isEmpty());
+            assertThat(handler.requestedBuffers.isEmpty()).isTrue();
```
```suggestion
            assertThat(handler.requestedBuffers).isEmpty();
```
There is `isEmpty()` method from assertJ

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriteRequestExecutorImplTest.java:
```diff
@@ -73,8 +78,8 @@ private void testCloseAfterSubmit(
         requestFun.accept(worker, request);
-        assertTrue(closingDeque.isEmpty());
-        assertFalse(request.isCancelled());
+        assertThat(closingDeque.isEmpty()).isTrue();
```
```suggestion
        assertThat(closingDeque).isEmpty();
```
Could be simplified

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriteRequestExecutorImplTest.java:
```diff
@@ -91,15 +96,15 @@ private void testSubmitFailure(
-            assertTrue(request.cancelled);
-            assertTrue(deque.isEmpty());
+            assertThat(request.cancelled).isTrue();
+            assertThat(deque.isEmpty()).isTrue();
```
```suggestion
            assertThat(deque).isEmpty();
```
Could be simplified

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriteRequestExecutorImplTest.java:
```diff
@@ -110,13 +115,13 @@ public void testCleanup() throws IOException {
-        assertTrue(deque.isEmpty());
-        assertTrue(request.isCancelled());
+        assertThat(requestProcessor.isStopped()).isTrue();
+        assertThat(deque.isEmpty()).isTrue();
```
```suggestion
        assertThat(deque).isEmpty();
```

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriterImplTest.java:
```diff
@@ -37,51 +37,58 @@
             result = writer.getAndRemoveWriteResult(CHECKPOINT_ID);
-            assertFalse(result.resultSubpartitionStateHandles.isDone());
-            assertFalse(result.inputChannelStateHandles.isDone());
+            assertThat(result.resultSubpartitionStateHandles.isDone()).isFalse();
+            assertThat(result.inputChannelStateHandles.isDone()).isFalse();
```
```suggestion
            assertThat(result.resultSubpartitionStateHandles).isNotDone();
            assertThat(result.inputChannelStateHandles).isNotDone();
```
assertJ supports some `CompletableFuture` methods

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriterImplTest.java:
```diff
@@ -37,51 +37,58 @@
         }
-        assertTrue(result.inputChannelStateHandles.isDone());
-        assertTrue(result.resultSubpartitionStateHandles.isDone());
+        assertThat(result.inputChannelStateHandles.isDone()).isTrue();
+        assertThat(result.resultSubpartitionStateHandles.isDone()).isTrue();
```
```suggestion
        assertThat(result.inputChannelStateHandles).isDone();
        assertThat(result.resultSubpartitionStateHandles).isDone();
```
assertJ supports some `CompletableFuture` methods

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriterImplTest.java:
```diff
@@ -135,145 +145,200 @@ public void testAbortOldAndStartNewCheckpoint() throws Exception {
                         fail("The result should have failed.");
                     } catch (Throwable throwable) {
-                        assertTrue(findThrowable(throwable, TestException.class).isPresent());
+                        assertThat(findThrowable(throwable, TestException.class).isPresent())
+                                .isTrue();
```
```suggestion
                        assertThat(findThrowable(throwable, TestException.class)).isPresent();
```
assertJ supports some `Optional` methods

1996fanrui replied:
Change to this:

```
                    assertThatThrownBy(() -> result42.getInputChannelStateHandles().get())
                            .as("The result should have failed.")
                            .hasCauseInstanceOf(TestException.class);
```

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/CheckpointInProgressRequestTest.java:
```diff
@@ -55,7 +55,7 @@ public void testNoCancelTwice() throws Exception {
         }
 
-        assertEquals(1, counter.get());
+        assertThat(counter.get()).isEqualTo(1);
```
```suggestion
        assertThat(counter.get()).isOne();
```

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/SequentialChannelStateReaderImplTest.java:
```diff
@@ -184,7 +186,7 @@ private Map<InputChannelInfo, List<Buffer>> collectBuffers(InputGate[] gates) th
             throws InterruptedException, java.util.concurrent.ExecutionException {
         for (InputGate gate : gates) {
-            assertTrue(gate.getStateConsumedFuture().isDone());
+            assertThat(gate.getStateConsumedFuture().isDone()).isTrue();
```
```suggestion
            assertThat(gate.getStateConsumedFuture()).isDone();
```
assertJ supports some `CompletableFuture` methods

snuyanzin commented:
In general looks good, i left some comments.
May be a dummy question to clarify: is there a way to determine which tests are related to unaligned checkpoint and which are not based e.g. on a package name?

1996fanrui commented:
Hi @snuyanzin , Thanks for your review. I have addressed all comments.

> Is there a way to determine which tests are related to unaligned checkpoint and which are not based e.g. on a package name?

Yes, I based package name. In fact, it is non-strict. The `flink-runtime` module is too large to be migrated in one go. So split some small packages to migrate.

reswqa commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/CheckpointInProgressRequestTest.java:
```diff
@@ -55,7 +55,7 @@ public void testNoCancelTwice() throws Exception {
         }
 
-        assertEquals(1, counter.get());
+        assertThat(counter.get()).isOne();
```
```suggestion
        assertThat(counter).hasValue(1);
```

RocMarshal commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateChunkReaderTest.java:
```diff
@@ -34,61 +34,67 @@
+                                                0);
+                            } finally {
+                                assertThat(serializer.failed).isTrue();
+                                assertThat(handler.requestedBuffers).isNotEmpty();
+                                assertThat(handler.requestedBuffers)
+                                        .allMatch(TestChannelStateByteBuffer::isRecycled);
```
```suggestion
                                assertThat(handler.requestedBuffers).isNotEmpty()
                                        .allMatch(TestChannelStateByteBuffer::isRecycled);
```

RocMarshal commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateChunkReaderTest.java:
```diff
@@ -34,61 +34,67 @@
-            assertTrue(
-                    handler.requestedBuffers.stream()
-                            .allMatch(TestChannelStateByteBuffer::isRecycled));
+            assertThat(handler.requestedBuffers).isNotEmpty();
+            assertThat(handler.requestedBuffers).allMatch(TestChannelStateByteBuffer::isRecycled);
```
```suggestion
            assertThat(handler.requestedBuffers)
                    .isNotEmpty()
                    .allMatch(TestChannelStateByteBuffer::isRecycled);
```

RocMarshal commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/InputChannelRecoveredStateHandlerTest.java:
```diff
@@ -27,23 +27,23 @@
     private static final int preAllocatedSegments = 3;
     private NetworkBufferPool networkBufferPool;
     private SingleInputGate inputGate;
     private InputChannelRecoveredStateHandler icsHander;
```
```suggestion
    private InputChannelRecoveredStateHandler icsHandler;
```

reswqa commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriterImplTest.java:
```diff
@@ -135,145 +144,196 @@ public void testAbortOldAndStartNewCheckpoint() throws Exception {
-                });
+    @Test
+    void testNoStartAfterClose() {
+        assertThatThrownBy(
+                        () ->
+                                unwrappingError(
+                                        IllegalStateException.class,
+                                        () -> {
+                                            ChannelStateWriterImpl writer = openWriter();
+                                            writer.close();
+                                            writer.start(
+                                                    42,
+                                                    CheckpointOptions
+                                                            .forCheckpointWithDefaultLocation());
+                                        }))
+                .isInstanceOf(IllegalStateException.class);
```
```suggestion
        Assertions.assertThatThrownBy(
                        () -> {
                            ChannelStateWriterImpl writer = openWriter();
                            writer.close();
                            writer.start(42, CheckpointOptions.forCheckpointWithDefaultLocation());
                        })
                .hasCauseInstanceOf(IllegalStateException.class);
```
Maybe we can using `hasCauseInstanceOf` directly instead of `unwrappingError`.

RocMarshal commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/RecordingChannelStateWriter.java:
```diff
@@ -32,8 +32,9 @@
 public class RecordingChannelStateWriter extends MockChannelStateWriter {
     private long lastStartedCheckpointId = -1;
     private long lastFinishedCheckpointId = -1;
```
Could this line be removed ?

RocMarshal reviewed (commented):
Cool~
Thanks for @1996fanrui contribution and review @snuyanzin .
I left a few comments. PTAL. thx.

reswqa reviewed (commented):
@1996fanrui Thanks for doing this migration, It looks very good overall. I have only a few suggestions.

1996fanrui commented:
Hi @RocMarshal @reswqa , thanks a lot for your review, I have updated.

reswqa reviewed (approved):
Thanks for the update, LGTM.

RocMarshal reviewed (approved):
thx for the update. LGTM~

1996fanrui commented:
Hi @XComp @snuyanzin , I have addressed all comments, please help take a look, thanks~

snuyanzin commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateCheckpointWriterTest.java:
```diff
@@ -48,20 +48,18 @@
     private final Random random = new Random();
 
-    @Rule public final TemporaryFolder temporaryFolder = new TemporaryFolder();
+    @TempDir private java.nio.file.Path temporaryFolder;
```
nit: 
```suggestion
    @TempDir private Path temporaryFolder;
```

1996fanrui replied:
Thanks for your review, updated.

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriterImplTest.java:
```diff
@@ -36,52 +35,57 @@
-        } finally {
-            assertTrue(dataBuf.isRecycled());
-        }
+        assertThatThrownBy(
```
we should make the `assertThatThrownBy` specific to the call that causes the issue.

1996fanrui replied:
The code after update, it should be specific. Thanks for your great suggestion, totally agree with your suggestion.

```
    @Test
    void testAddEventBuffer() throws Exception {

        NetworkBuffer dataBuf = getBuffer();
        NetworkBuffer eventBuf = getBuffer();
        eventBuf.setDataType(Buffer.DataType.EVENT_BUFFER);

        runWithSyncWorker(
                (writer, worker) -> {
                    callStart(writer);
                    callAddInputData(writer, dataBuf);
                    worker.processAllRequests();

                    callAddInputData(writer, eventBuf);
                    assertThatThrownBy(worker::processAllRequests)
                            .isInstanceOf(IllegalArgumentException.class);
                });
        assertThat(dataBuf.isRecycled()).isTrue();
    }
```

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriterImplTest.java:
```diff
@@ -135,154 +142,165 @@ public void testAbortOldAndStartNewCheckpoint() throws Exception {
-                    }
-                });
+    @Test
+    void testBuffersRecycledOnError() {
+        assertThatThrownBy(
+                        () -> {
+                            NetworkBuffer buffer = getBuffer();
+                            try (ChannelStateWriterImpl writer =
+                                    new ChannelStateWriterImpl(
+                                            TASK_NAME,
+                                            new ConcurrentHashMap<>(),
+                                            failingWorker(),
+                                            5)) {
+                                writer.open();
+                                callAddInputData(writer, buffer);
+                            } finally {
+                                assertThat(buffer.isRecycled()).isTrue();
+                            }
+                        })
+                .hasCauseInstanceOf(TestException.class);
```
```suggestion
    void testBuffersRecycledOnError() throws IOException {
        NetworkBuffer buffer = getBuffer();
        try (ChannelStateWriterImpl writer =
                new ChannelStateWriterImpl(
                        TASK_NAME, new ConcurrentHashMap<>(), failingWorker(), 5)) {
            writer.open();
            assertThatThrownBy(() -> callAddInputData(writer, buffer))
                    .isInstanceOf(RuntimeException.class)
                    .hasCauseInstanceOf(TestException.class);
            assertThat(buffer.isRecycled()).isTrue();
        }
```
We can assert on the exception here in a more-specific manner as well. This also improves the readability of the test.

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriterImplTest.java:
```diff
@@ -135,154 +142,165 @@ public void testAbortOldAndStartNewCheckpoint() throws Exception {
-        writer.open();
-        worker.setThrown(new TestException());
-        unwrappingError(TestException.class, () -> callStart(writer));
+        assertThatThrownBy(
+                        () -> {
+                            writer.open();
+                            worker.setThrown(new TestException());
+                            callStart(writer);
+                        })
+                .hasCauseInstanceOf(TestException.class);
```
```suggestion
        writer.open();
        worker.setThrown(new TestException());
        assertThatThrownBy(() -> callStart(writer)).hasCauseInstanceOf(TestException.class);
```
More-specific `assertThatThrownBy` help here as well.

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriterImplTest.java:
```diff
@@ -135,154 +142,165 @@ public void testAbortOldAndStartNewCheckpoint() throws Exception {
-            }
-            writer.start(maxCheckpoints, CheckpointOptions.forCheckpointWithDefaultLocation());
-        }
+        assertThatThrownBy(
+                        () -> {
+                            try (ChannelStateWriterImpl writer =
+                                    new ChannelStateWriterImpl(
+                                            TASK_NAME,
+                                            0,
+                                            getStreamFactoryFactory(),
+                                            maxCheckpoints)) {
+                                writer.open();
+                                for (int i = 0; i < maxCheckpoints; i++) {
+                                    writer.start(
+                                            i,
+                                            CheckpointOptions.forCheckpointWithDefaultLocation());
+                                }
+                                writer.start(
+                                        maxCheckpoints,
+                                        CheckpointOptions.forCheckpointWithDefaultLocation());
+                            }
+                        })
+                .isInstanceOf(IllegalStateException.class);
```
```suggestion
        try (ChannelStateWriterImpl writer =
                new ChannelStateWriterImpl(
                        TASK_NAME, 0, getStreamFactoryFactory(), maxCheckpoints)) {
            writer.open();
            for (int i = 0; i < maxCheckpoints; i++) {
                writer.start(i, CheckpointOptions.forCheckpointWithDefaultLocation());
            }

            assertThatThrownBy(
                            () ->
                                    writer.start(
                                            maxCheckpoints,
                                            CheckpointOptions.forCheckpointWithDefaultLocation()))
                    .isInstanceOf(IllegalStateException.class);
        }
```

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriterImplTest.java:
```diff
@@ -135,154 +142,165 @@ public void testAbortOldAndStartNewCheckpoint() throws Exception {
-                    writer.start(42, CheckpointOptions.forCheckpointWithDefaultLocation());
-                });
+    @Test
+    void testStartNotOpened() {
+        assertThatThrownBy(
+                        () -> {
+                            try (ChannelStateWriterImpl writer =
+                                    new ChannelStateWriterImpl(
+                                            TASK_NAME, 0, getStreamFactoryFactory())) {
+                                callStart(writer);
+                            }
+                        })
+                .hasCauseInstanceOf(IllegalStateException.class);
```
```suggestion
        void testStartNotOpened() throws IOException {
        try (ChannelStateWriterImpl writer =
                new ChannelStateWriterImpl(TASK_NAME, 0, getStreamFactoryFactory())) {
            assertThatThrownBy(() -> callStart(writer))
                    .hasCauseInstanceOf(IllegalStateException.class);
        }
```

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriterImplTest.java:
```diff
@@ -135,154 +142,165 @@ public void testAbortOldAndStartNewCheckpoint() throws Exception {
-                    callAddInputData(writer);
-                });
+    @Test
+    void testNoStartAfterClose() {
+        assertThatThrownBy(
+                        () -> {
+                            ChannelStateWriterImpl writer = openWriter();
+                            writer.close();
+                            writer.start(42, CheckpointOptions.forCheckpointWithDefaultLocation());
+                        })
+                .hasCauseInstanceOf(IllegalStateException.class);
```
```suggestion
        void testNoStartAfterClose() throws IOException {
        ChannelStateWriterImpl writer = openWriter();
        writer.close();
        assertThatThrownBy(
                        () ->
                                writer.start(
                                        42, CheckpointOptions.forCheckpointWithDefaultLocation()))
                .hasCauseInstanceOf(IllegalStateException.class);
```

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriterImplTest.java:
```diff
@@ -135,154 +142,165 @@ public void testAbortOldAndStartNewCheckpoint() throws Exception {
-            throw findThrowable(e, clazz).map(te -> (Exception) te).orElse(e);
-        }
+    @Test
+    void testNoAddDataAfterClose() {
+        assertThatThrownBy(
+                        () -> {
+                            ChannelStateWriterImpl writer = openWriter();
+                            callStart(writer);
+                            writer.close();
+                            callAddInputData(writer);
+                        })
```
```suggestion
    void testNoAddDataAfterClose() throws IOException {
        ChannelStateWriterImpl writer = openWriter();
        callStart(writer);
        writer.close();
        assertThatThrownBy(() -> callAddInputData(writer))
```

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/SequentialChannelStateReaderImplTest.java:
```diff
@@ -65,49 +68,48 @@
-            {"ReadPermutedStateWithReducedBuffer", 10, 10, 10, 20, 10},
-            {"ReadPermutedStateWithIncreasedBuffer", 10, 10, 10, 10, 20},
-        };
+    @Parameters
```
```suggestion
    @Parameters(name =
                    "{0}: stateParLevel={1}, statePartsPerChannel={2}, stateBytesPerPart={3},  parLevel={4}, bufferSize={5}")
```
The JUnit5 `@Parameters` annotation also supports custom display names

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriteRequestDispatcherTest.java:
```diff
@@ -42,56 +44,66 @@
-    private final Optional<Class<?>> expectedException;
-    public static final long CHECKPOINT_ID = 42L;
-
     @Parameters
```
```suggestion
    @Parameters(name = "expectedException={0} requests={1}")
```

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/SequentialChannelStateReaderImplTest.java:
```diff
@@ -65,49 +68,48 @@
-        };
+    @Parameters
+    public static List<Object[]> parameters() {
+        ArrayList<Object[]> params = new ArrayList<>();
+        params.add(new Object[] {"NoStateAndNoChannels", 0, 0, 0, 0, 0});
+        params.add(new Object[] {"NoState", 0, 10, 10, 10, 10});
+        params.add(new Object[] {"ReadPermutedStateWithEqualBuffer", 10, 10, 10, 10, 10});
+        params.add(new Object[] {"ReadPermutedStateWithReducedBuffer", 10, 10, 10, 20, 10});
+        params.add(new Object[] {"ReadPermutedStateWithIncreasedBuffer", 10, 10, 10, 10, 20});
+        return params;
```
```suggestion
        return Arrays.asList(
                new Object[] {"NoStateAndNoChannels", 0, 0, 0, 0, 0},
                new Object[] {"NoState", 0, 10, 10, 10, 10},
                new Object[] {"ReadPermutedStateWithEqualBuffer", 10, 10, 10, 10, 10},
                new Object[] {"ReadPermutedStateWithReducedBuffer", 10, 10, 10, 20, 10},
                new Object[] {"ReadPermutedStateWithIncreasedBuffer", 10, 10, 10, 10, 20});
```
nit: you don't have to change this if you don't agree.

XComp reviewed (changes requested):
Thanks @1996fanrui for this PR and @snuyanzin , @RocMarshal and @reswqa for reviewing this big change thoroughly. I did a pass over it and had some final comments.

A final thought on collecting expected exceptions. We should really be specific in the callback of an `assertThatThrownBy` call to be sure that the exception is actually thrown in the expected place. I understand that this wasn't the case in the JUnit4 version either. But I think that it's worth it to add this improvement as part of the JUnit5 migration.

reswqa commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateChunkReaderTest.java:
```diff
@@ -34,61 +34,68 @@
-                    handler.requestedBuffers.stream()
-                            .allMatch(TestChannelStateByteBuffer::isRecycled));
-        }
+        assertThatThrownBy(
+                        () -> {
+                            try (FSDataInputStream stream = getStream(serializer, 10)) {
+                                new ChannelStateChunkReader(serializer)
+                                        .readChunk(
+                                                stream,
+                                                serializer.getHeaderLength(),
+                                                handler,
+                                                "channelInfo",
+                                                0);
+                            } finally {
+                                assertThat(serializer.failed).isTrue();
+                                assertThat(handler.requestedBuffers)
+                                        .isNotEmpty()
+                                        .allMatch(TestChannelStateByteBuffer::isRecycled);
+                            }
+                        })
+                .isInstanceOf(TestException.class);
```
I like the idea that the assert's callback should be as small as possible to avoid asserting other code locations by accident. Thanks @XComp for pointing this out, big +1 for this!

1996fanrui commented:
Hi @XComp , thanks for your hard review. I have addressed all comments.

And thanks @XComp  @snuyanzin @reswqa and @RocMarshal again, I learned a lot in this PR. It will be useful for my future contributions.

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriterImplTest.java:
```diff
@@ -36,52 +35,49 @@
-            assertTrue(dataBuf.isRecycled());
-        }
+
+        runWithSyncWorker(
```
I also noticed the following: Refactoring this code in a way that we're using `assertThatThrownBy` is used changes the behavior of `runWithSyncWorker` slightly because of `worker.processAllRequests()` in [ChannelStateWriterImplTest:313](https://github.com/apache/flink/blob/772de6fc8698eea8b7b42c3a1de4282a61ba7e8f/flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriterImplTest.java#L313). The old implementation didn't call `processAllRequests()` if an exception was thrown. The refactored version would actually cause this method to be called which might change the behavior. You might want to double-check whether that's something we want.

1996fanrui replied:
I checked all caller. All the unit tests of `@Test(expected = Exception.class)` throw an exception in the last line of runWithSyncWorker or it just has one line, so I think the old and new codes are the same.

XComp replied:
I guess, you misunderstood me here: `worker.processAllRequests()` in [ChannelStateWriterImplTest:319](https://github.com/apache/flink/blob/5596204e67027ff924b97616b5793c8073434412/flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriterImplTest.java#L319) is called after the callback function is executed. In the old code, we threw an Exception within the callback function which resulted in the command in line 319 not being called. In the new version of the code, we rely on `assertThatThrownBy` which doesn't throw an Exception and, therefore, line 319 is executed, i.e. `worker.processAllRequests()` is called in the expected case. This feels like an unwanted change? :thinking:

1996fanrui replied:
When the callback doesn't throw exception. Old and new code is same.

Some old code tests rely on `runWithSyncWorker -> worker.processAllRequests()` to throw exceptions for tests that expect Exception. 

In the new code, I have additionally called worker.processAllRequests() in the callback, so `runWithSyncWorker -> worker.processAllRequests()` will not process any requests.

Please correct me, if anything is wrong. 

Or if we worry about which tests are wrong, we can refactor them into non-specific checks, what do you think?

XComp replied:
To be on the safe side, we should remove `worker.processAllRequests()` in [ChannelStateWriterImplTest:316](https://github.com/apache/flink/blob/3156ff6219452c51c010ded37197d01c1c5924c1/flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriterImplTest.java#L316) and move it into the tests where this call is still required. This is test-specific code (for tests where no exception is expected) rather than generic code. WDYT?

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriterImplTest.java:
```diff
@@ -135,100 +134,98 @@ public void testAbortOldAndStartNewCheckpoint() throws Exception {
+    }
+
+    @Test
+    void testNoAddDataAfterFinished() {
```
This entire change block has examples where we are too broad in asserting the exception.

1996fanrui replied:
After updated, I still call `processAllRequests` twice to ensure the catch is specific.

It's similar to https://github.com/apache/flink/pull/21368#discussion_r1037196683 , the first one is normal, the second one should throw exception.

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriterImplTest.java:
```diff
@@ -237,52 +234,41 @@ public void testLimit() throws IOException {
+        ChannelStateWriterImpl writer = openWriter();
+        callStart(writer);
+        writer.close();
+        assertThatThrownBy(() -> callAddInputData(writer));
```
We're missing the expected exception here

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriterImplTest.java:
```diff
@@ -36,52 +35,49 @@
 
 /** {@link ChannelStateWriterImpl} lifecycle tests. */
-public class ChannelStateWriterImplTest {
+class ChannelStateWriterImplTest {
```
nit: I cannot comment on that line because it wasn't touched in this PR but we could also change the signature of `runWithSyncWorker(BiConsumerWithException<ChannelStateWriter, SyncChannelStateWriteRequestExecutor, Exception> testFn)` to `runWithSyncWorker(BiConsumerWithException<ChannelStateWriter, SyncChannelStateWriteRequestExecutor> testFn)` since the exception throwing is not utilized anymore when relying on a more specific `assertThatThrownBy` in each test.

1996fanrui replied:
After tried, I found we cannot change it.

Sometimes, `testFn` calls the ` worker.processAllRequests()`, and ` worker.processAllRequests()` won't throw exception, so `testFn` cannot be changed from `BiConsumerWithException` to `BiConsumer`.

XComp replied:
ok, fair point. :+1:

XComp reviewed (commented):
Thanks for addressing my comments, @1996fanrui . I went over the changes once more. Please find more comments below.

1996fanrui commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriterImplTest.java:
```diff
@@ -36,52 +35,49 @@
+        runWithSyncWorker(
+                (writer, worker) -> {
+                    callStart(writer);
+                    callAddInputData(writer, dataBuf);
```
Yes , I'm sure. 

Channel state just allows the data buffer, and doesn't allow the event buffer. This unit-test checks whether it throw `IllegalArgumentException` when adding an event buffer. 

So I call `processAllRequests` twice, the first one, it's a `start request` and `add data buffer request`, the second one, it's a `add event buffer request`. 

The first one is normal, and the second one should throw exception. It's more specific.

XComp replied:
I guess, you're right. But the test is slightly changed by you adding `dataBuf` before adding `eventBuf` because we're checking whether `dataBuf` is recycled afterwards. The semantics of this check are changed due to your change: Previously, the test checked that a buffer was recycled which should have been processed but wasn't processed because of the Exception the `eventBuf` caused. Now, we check that the `dataBuf` is recycled after being processed properly. I'm curious about your opinion here.

1996fanrui replied:
> he test checked that a buffer was recycled which should have been processed but wasn't processed because of the Exception the eventBuf caused. 

Good catch, I didn't realize it before. I just thought that test wanted to verify whether the data buffer is normal.

Now I understand it, thanks for your reminder~

XComp replied:
Thanks for the quick response, @1996fanrui. Looking at this specific test, we will see now that we run into the same issue which I explained in [my other comment](https://github.com/apache/flink/pull/21368#discussion_r1039276265). The expected `IllegalArgumentException` is caught and handled by `assertThatThrownBy`. As a consequence, the call chain continues within `runWithSyncWorker` making `processAllRequests` being called once more which might affect `dataBuf` in this specific case. I'm gonna propose a way to deal with this in the other thread to keep that discussion there.

XComp commented:
I replied to your comments, @1996fanrui. I'm looking forward to your response.

1996fanrui commented:
> I replied to your comments, @1996fanrui. I'm looking forward to your response.

@XComp , thanks for your hard review, update~

XComp reviewed (approved):
Thanks for addressing my comments. The PR looks good. :+1:

XComp commented:
@rkhachatryan do you want to have a brief look over the changes to make sure that we didn't accidentally change some test logic? Otherwise, I would go ahead and merge it tomorrow after CI is green

1996fanrui commented:
> Thanks for addressing my comments. The PR looks good. 👍
> 
> @1996fanrui just as hint for future PRs. It helps to add new commits while being in the review process and only do a squash after you're sure that the review is done (usually that can be even done by the committer when merging the PR). Github provides functionality to check the diff after a forced-push. But having separate commits also makes the incremental changes available locally when reviewing the PR using an IDE.

Thanks for your reminder.

1996fanrui commented:
@flinkbot run azure

XComp commented:
Why another review. I was just waiting with merging because I wanted to give @rkhachatryan the opportunity to verify that we didn't miss something. Merging the PR is on my list of things to do for Monday. :-)

rkhachatryan commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriterImplTest.java:
```diff
@@ -135,100 +130,99 @@ public void testAbortOldAndStartNewCheckpoint() throws Exception {
+        assertThatThrownBy(
+                        () ->
+                                executeCallbackAndProcessWithSyncWorker(
+                                        (Consumer<ChannelStateWriter>) this::callAddInputData))
```
nit: the cast here seems unnecessary

rkhachatryan reviewed (commented):
Thanks for the PR and for pulling me in.
I've left only some minor comments, PTAL.

1996fanrui commented:
> Thanks for the PR and for pulling me in. I've left only some minor comments, PTAL.

Hi @rkhachatryan , thanks for your review, I have updated.

XComp reviewed (approved):
Thanks for giving it another pass @rkhachatryan . I'm gonna go ahead and merge it as soon as CI is green.

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateChunkReaderTest.java:
```diff
@@ -34,61 +34,66 @@
+                                                    "channelInfo",
+                                                    0))
+                    .isInstanceOf(TestException.class);
+            assertThat(serializer.failed).isTrue();
```
Fair enough. Another approach to describe the test scenario more specifically would be adding a proper assert message describing what and why the condition that is subject to the assert needs to be met.

rkhachatryan reviewed (approved):
Thanks for updating the PR, LGTM

1996fanrui commented:
@flinkbot run azure

XComp commented:
CI is green. About the rerun: It would be also nice to investigate build failures. The build failures we've seen in this PR's CI can be explained with FLINK-29427 and FLINK-30355. The latter one seems to be an issue that we haven't observed so far (based on a Jira search). It would have been a pity if we would have missed that one.

Such an analysis is also good to determine whether the build failures are related to the PR's changes or not. In this case, it's obvious that they are not because we only refactored test classes. Therefore, we're good to go. I will merge the PR. Thanks for your contribution @1996fanrui and everyone else for their reviews.

1996fanrui commented:
Thanks everyone who reviewed 👍🏻
